File: src/breslin/clientside/game/partido/gamePartido.cpp
//header
#include "gamePartido.h"

//log
#include "../../tdreamsock/dreamSockLog.h"

//ObjectTitle
#include "../../billboard/objectTitle.h"

//byteBuffer
#include "../../bytebuffer/byteBuffer.h"

//shape
#include "../../shape/shape.h"

//application
#include "../../application/partido/applicationPartido.h"

//states
#include "states/gamePartidoStates.h"

/***************************************
*			          CONSTRUCTORS
***************************************/
GamePartido::GamePartido(ApplicationPartido* applicationPartido) : Game(applicationPartido)
{
	mApplicationPartido = applicationPartido;

/*
  //reset battle vars
        mBattleStart = false;
        mCorrectAnswerStart = false;
        mApplicationPartido->mGameReset = false;

        //reset correctAnswer vars
        mCorrectAnswerStart = false;
        mCorrectAnswer      = false;
        mCorrectAnswerEnd   = false;

        //answer time
        mApplicationPartido->mAnswerTime = 0;

*/


	//battle
	mBattleStart = false;
	mFirstTimeExecute = true;

	//correctAnswer
	mCorrectAnswerStart = false;
	mCorrectAnswerEnd = false;

	//states
        mPartidoStateMachine =  new StateMachine<GamePartido>(this);
        mPartidoStateMachine->setCurrentState      (PLAY_PARTIDO_GAME::Instance());
        mPartidoStateMachine->setPreviousState     (PLAY_PARTIDO_GAME::Instance());
        mPartidoStateMachine->setGlobalState       (NULL);

	//states
        mBattleStateMachine =  new StateMachine<GamePartido>(this);
        mBattleStateMachine->setCurrentState      (BATTLE_OFF::Instance());
        mBattleStateMachine->setPreviousState     (BATTLE_OFF::Instance());
        mBattleStateMachine->setGlobalState       (NULL);

	mApplicationPartido->createBattleScreen();
        mApplicationPartido->hideBattleScreen();
        mApplicationPartido->createCorrectAnswerScreen();
        mApplicationPartido->hideCorrectAnswerScreen();
}

GamePartido::~GamePartido()
{
}

void GamePartido::processUpdate()
{
	Game::processUpdate();
        mPartidoStateMachine->update();
        mBattleStateMachine->update();
}

void GamePartido::processMoveControls()
{
	if (mBattleStateMachine->currentState() != BATTLE_OFF::Instance())
	{
		//we are in battle			
	}	
	else
	{
		//not in battle let player move
		Game::processMoveControls();
	}
}

void GamePartido::checkByteBuffer(ByteBuffer* byteBuffer)
{
	Game::checkByteBuffer(byteBuffer);

        byteBuffer->BeginReading();

        int type = byteBuffer->ReadByte();
       
	switch(type)
        {
                case mMessageAskQuestion:
                        askQuestion(byteBuffer);
                        break;

                case mMessageBattleStart:
			mBattleStart = true;
                        break;

                case mMessageBattleEnd:
 			mBattleStateMachine->changeState(BATTLE_OFF::Instance());                        
			break;
                
		case mMessageCorrectAnswerStart:
			mCorrectAnswerStart = true;
                        correctAnswer(byteBuffer);
                        break;
		
		case mMessageCorrectAnswerEnd:
			mCorrectAnswerEnd = true;
                        break;
        }
}

void GamePartido::correctAnswer(ByteBuffer* byteBuffer)
{
        mApplicationPartido->mStringCorrectAnswer.clear();

        int length = byteBuffer->ReadByte();
        for (int i = 0; i < length; i++)
        {
                char c =  byteBuffer->ReadByte();
                mApplicationPartido->mStringCorrectAnswer.append(1,c);
        }
        if (mApplicationPartido->mLabelCorrectAnswer && mApplicationPartido->mStringCorrectAnswer.size() > 0)
        {
                mApplicationPartido->mLabelCorrectAnswer->setCaption(mApplicationPartido->mStringCorrectAnswer);
        }
        else
        {
                LogString("no label or no string");
        }
}

void GamePartido::askQuestion(ByteBuffer* byteBuffer)
{
	mApplicationPartido->mStringQuestion.clear();
        int length = byteBuffer->ReadByte();
        for (int i = 0; i < length; i++)
        {
        	char c =  byteBuffer->ReadByte();
                mApplicationPartido->mStringQuestion.append(1,c);
        }

	const char * q = mApplicationPartido->mStringQuestion.c_str();
        //LogString("q:%s",q);

	if (!mApplicationPartido->mLabelQuestion)
	{
		LogString("GamePartido::askQuestion : no mLabelQuestion exists yet.");
	}
	if (mApplicationPartido->mStringQuestion.size() < 1)
	{
		LogString("GamePartido::askQuestion : mStringQuestion less than 1 in size.");
	}
	if (mApplicationPartido->mLabelQuestion && mApplicationPartido->mStringQuestion.size() > 0)
	{
		mApplicationPartido->mLabelQuestion->setCaption(mApplicationPartido->mStringQuestion);
		//LogString("just setCaption on mLabelQuestion");
	}
	else
	{
		LogString("GamePartido::askQuestion : no label or no string");
	}
	
	//reset mAnswerTime	
	mApplicationPartido->mAnswerTime = 0; 
}

void GamePartido::reset()
{
	Game::reset();
      	
	//hide screens 
	mApplicationPartido->hideBattleScreen();
	mApplicationPartido->hideCorrectAnswerScreen();
 	
	//reset text boxes for battle ...actually let's not
        mApplicationPartido->mStringAnswer.clear();
        mApplicationPartido->mLabelQuestion->setCaption("");
        mApplicationPartido->mLabelAnswer->setCaption("");

	//reset text boxes for showCorrectAnswer ...actually let's not
        mApplicationPartido->mLabelCorrectAnswer->setCaption("");
        mApplicationPartido->mStringCorrectAnswer.clear();

	//reset battle vars
        mBattleStart = false;
        mCorrectAnswerStart = false;
        mApplicationPartido->mGameReset = false;

	//reset correctAnswer vars
        mCorrectAnswerStart = false;
        mCorrectAnswerEnd   = false;

	//answer time
        mApplicationPartido->mAnswerTime = 0;
}

File: src/breslin/clientside/game/partido/game_partido.php
var GamePartido = new Class(
{

Extends: Game,

initialize: function(applicationPartido)
{
	this.parent(applicationPartido);

	this.mApplicationPartido = applicationPartido;
	
	//questions
	this.mMessageAskQuestion    = -76;

	//show correct answer
	this.mMessageShowCorrectAnswer    = -62;
		
	//battle
	this.mMessageBattleStart = -75;
	this.mMessageBattleEnd   = -74;

	this.mBattleStart = false;

	//correctAnswer
	this.mCorrectAnswerStart = false;
	this.mCorrectAnswerEnd   = false;

	//partido state machines
	this.mPartidoStateMachine = new StateMachine(this);

        this.mPLAY_PARTIDO_GAME              = new PLAY_PARTIDO_GAME(this);
 	this.mRESET_PARTIDO_GAME             = new RESET_PARTIDO_GAME(this);
	
	this.mPartidoStateMachine.changeState(this.mPLAY_PARTIDO_GAME);
 
	//battle state machine	
	this.mBattleStateMachine  = new StateMachine(this);

	this.mBATTLE_OFF                     = new BATTLE_OFF(this);
        this.mANSWER_QUESTION                = new ANSWER_QUESTION(this);
        this.mSHOWCORRECTANSWER_PARTIDO_GAME = new SHOWCORRECTANSWER_PARTIDO_GAME(this);
        
        this.mBattleStateMachine.changeState(this.mBATTLE_OFF);
},

log: function(msg)
{
	setTimeout(function()
        {
        	throw new Error(msg);
        }, 0);
},


askQuestion: function(byteBuffer)
{
        this.mApplicationPartido.mStringQuestion = '';

        this.mApplicationPartido.mStringQuestion = byteBuffer.readByte();

	if (this.mApplicationPartido.mLabelQuestion)
	{
		this.mApplicationPartido.mLabelQuestion.value = this.mApplicationPartido.mStringQuestion; 
	}
	else
	{
	}
       	
	//reset mAnswerTime 
	this.mApplicationPartido.mAnswerTime = 0;
},

correctAnswer: function(byteBuffer)
{
	this.log('correctAnswer called');
        this.mApplicationPartido.mStringCorrectAnswer = '';

        this.mApplicationPartido.mStringCorrectAnswer = byteBuffer.readByte();
	this.log('mStringCorrectAnswer:' + this.mApplicationPartido.mStringCorrectAnswer);

        if (this.mApplicationPartido.mLabelCorrectAnswer)
        {
                this.mApplicationPartido.mLabelCorrectAnswer.value = this.mApplicationPartido.mStringCorrectAnswer;
        }
        else
        {
        }
},

reset: function()
{
	this.parent();
	//hide screens
        this.mApplicationPartido.hideBattleScreen();
        this.mApplicationPartido.hideCorrectAnswerScreen();

        //reset text boxes for battle ...actually let's not
	/*
        mApplicationPartido->mStringAnswer.clear();
        mApplicationPartido->mLabelQuestion->setCaption("");
        mApplicationPartido->mLabelAnswer->setCaption("");

        //reset text boxes for showCorrectAnswer ...actually let's not
        mApplicationPartido->mLabelCorrectAnswer->setCaption("");
        mApplicationPartido->mStringCorrectAnswer.clear();
*/
        //reset battle vars
        this.mBattleStart = false;
        this.mCorrectAnswerStart = false;
        this.mApplicationPartido.mGameReset = false;

        //reset correctAnswer vars
        this.mCorrectAnswerStart = false;
        this.mCorrectAnswer      = false;
        this.mCorrectAnswerEnd   = false;

        //answer time
        this.mApplicationPartido.mAnswerTime = 0;

},

processUpdate: function()
{
	this.parent();
        this.mPartidoStateMachine.update();
        this.mBattleStateMachine.update();
}


});
